core/trie/src/node_header.rs

```diff
@@ -16,62 +16,105 @@
 
 //! The node header.
 
+use crate::trie_constants;
 use codec::{Encode, Decode, Input, Output};
-use super::{EMPTY_TRIE, LEAF_NODE_OFFSET, LEAF_NODE_BIG, EXTENSION_NODE_OFFSET,
-	EXTENSION_NODE_BIG, BRANCH_NODE_NO_VALUE, BRANCH_NODE_WITH_VALUE, LEAF_NODE_THRESHOLD,
-	EXTENSION_NODE_THRESHOLD, LEAF_NODE_SMALL_MAX, EXTENSION_NODE_SMALL_MAX};
+use rstd::iter::once;
 
-/// A node header.
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub enum NodeHeader {
+/// A node header
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub(crate) enum NodeHeader {
 	Null,
-	Branch(bool),
-	Extension(usize),
+	Branch(bool, usize),
 	Leaf(usize),
 }
 
+/// NodeHeader without content
+pub(crate) enum NodeKind {
+	Leaf,
+	BranchNoValue,
+	BranchWithValue,
+}
+
 impl Encode for NodeHeader {
```

Comment:
Isn't it the same encoding/decoding as `ReferenceTrieStreamNoExt`? Can't we just re-use that?

Reply:
That is the same implementation, we got this double implementation situation (reference-test from trie is also problematic for other reason).
It could be possible to move this code in its own crate in trie, but it will not be in substrate.
We could also use this implementation from reference trie but at the cost of a dependency to substrate (maybe with a published crate this won't be such an issue).

Reply:
Thinking twice, codec implementation needs to be in substrate repo, so to use it for tests in trie repo we could create a new crate core/trie/nodecodec that does not depends on substrate crate and could easily be published.
That would not solve the issue with trie repo reference trie crate dependency but would remove this duplicated code.
Do you think it shall be done now or later (with a issue of course)?

Reply:
Ok, I guess it is fine as it is